mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+    let gb_range_start = PageAddr::new(SupervisorPhysAddr::supervisor(
+        PageSize::Size1G.round_down(phys_addr.bits()),
+    ))
+    .unwrap();
+    for i in 0u64..=15 {
+        // Unwrap OK: We are starting with an aligned address and are adding 0 or 16K 4K pages (64MB)
+        let range_addr = gb_range_start.checked_add_pages(i * 16 * 1024).unwrap();
+        // Sanity check
+        let _ = get_l2_entry(range_addr);
+        match memory_type {
+            MttMemoryType::NonConfidential => set_l2_entry(range_addr, MttNonConfidential1G),
+            MttMemoryType::Confidential => set_l2_entry(range_addr, MttConfidential1G),
+        }
+    }
+}
+
+// Changes the L2 type of each of the 16 consecutive entries for the the 1GB range containing the address
+// from 1GB to 2MBx32MB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential)
+fn fracture_1gb_range(phys_addr: SupervisorPageAddr) {
+    // Unwrap OK: phys_addr is aligned on a 4K boundary, and we are getting the 1GB start address containing the page
+    let gb_range_start = PageAddr::new(SupervisorPhysAddr::supervisor(
+        PageSize::Size1G.round_down(phys_addr.bits()),
+    ))
+    .unwrap();
+    let first_entry_type = get_l2_entry(gb_range_start);
+    // Enforce the invariant that all 16 L2 entries for the 1GB range must have the same type
+    for i in 0u64..=15 {
+        // Unwrap OK: We are starting with an aligned address and are adding 0 or 16K 4K pages (64MB)
+        let range_addr = gb_range_start.checked_add_pages(i * 16 * 1024).unwrap();
+        let mtt_entry = get_l2_entry(range_addr);
+        match mtt_entry {
+            MttNonConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(range_addr, MttL2EntryType::non_confidential_64_mb_range())
+            }
+            MttConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(range_addr, MttL2EntryType::confidential_64_mb_range())
+            }
+            _ => {
+                panic!(
+                    "fracture_1gb_range() error: start: {:x}, current: {:x} first_entry: {first_entry_type:?} current_entry: {mtt_entry:?}",
+                    gb_range_start.bits(), range_addr.bits()
+                )
+            }
+        }
+    }
+}
+
+// Changes the L2 type of 64MB range containing the address to 4Kx16KB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential).
+fn fracture_64mb_range(phys_addr: SupervisorPageAddr) {
+    let mtt_entry = get_l2_entry(phys_addr);
+    let mtt_l1_page = get_l1_page(phys_addr);
+    let l1_page_addr = mtt_l1_page.bits();
+    match mtt_entry {
+        Mtt2MBPages(value) => {
+            // Mark the entire 64MB as non-confidential at the outset
+            // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
```

Comment:
This is not is enforced here, the function takes _any_ page's address.

Reply:
Seems like the input parameter (`phys_addr`) is only used to compute the corresponding L2 entry. Since the MTT is accessible only to Salus, and the L1 pages are pre-allocated, are we violating any safety guarantees? Note that this raises the question of how we can trust platform parameters like the MTTP and L1 page pool, but that applies to other structures like the DT, etc. as well.

Reply:
Sorry, I misread this, and thought the write was going to and address offset from phys_addr.

What would make this clear is a MTT structure that owns the MTT pages and exposes a safe method that updates them for a given address. Then there will be one, central unsafe pointer deref and the currently scattered writes will use the safe API instead. Can that work?

Reply:
The latest refactoring uses something similar to the above suggestion...PTAL.